### PR TITLE
Updates Models to support new dataloader format for lists (__values and __offsets in dict) and scalar (1D)

### PR DESCRIPTION
**Pull request by gabrielspmoreira**
Fixes [#819](https://github.com/NVIDIA-Merlin/Merlin/issues/819)
It also resolves the [Capture shapes everywhere #823](https://github.com/NVIDIA-Merlin/Merlin/issues/823) for the Merlin Models library

### Goals :soccer:
This PR changes Merlin Models to support dataloader changes from [PR #101](https://github.com/NVIDIA-Merlin/dataloader/pull/101):
- List features are now represented by two dict keys with the following format: `{feature_name}__values`, `{feature_name}__offsets`, instead of a tuple
- Scalar features are now 1D tensors, instead of 2D

It also updates code from using `ColumnSchema` `"value_count"` to use the new ColumnSchema `shape` API.

### Implementation Details :construction:
Merlin Models had code in many places that converts the tuple representation into RaggedTensor, SparseTensor, or Tensor.
This PR centralizes the conversion from the dataloader representation of list features and eliminate duplicate code on lists conversion.

- Created centralized blocks to prepare scalar and list features (`PrepareFeatures`, `PrepareListFeatures`)
- Removed duplicated blocks to manage lists (`ListToDense`, `ListToRagged`, `ListToSparse`)
- Replaced usage of "properties/value_count" to use the new `shape`

*Other*
- Changed NVT preproc workflow of `examples/usecases/transformers-next-item-prediction.ipynb` to fix tagging and make it simpler

### Testing Details :mag:
- Fixed dozens of unit and integration tests to match the new data loader and shape API

### CODE REVIEW

File: merlin/models/tf/inputs/base.py [hunk 1]
```diff
@@ -208,7 +200,13 @@ def InputBlock(
             name="continuous_projection",
         )
 
-    return ParallelBlock(branches, aggregation=aggregation, post=post, is_input=True, **kwargs)
+    _pre = PrepareFeatures(schema)
```
Comment (oliverholworthy) [inline merlin/models/tf/inputs/base.py:203]: does every model have PrepareFeatures as the first layer? if so, maybe we don't need this in the InputBlock?

**Comment by oliverholworthy** [comment]
Re: issue with loader returning SparseTensor's. I think this is because in some cases we're now setting the value_count.max property (specifying the shape of ragged dimensions) which wasn't previously set. Until we remove the sparse output option in the dataloader https://github.com/NVIDIA-Merlin/dataloader/pull/103 . One intermediate step could be to update `_augment_schema`  helper function to always remove `value_count` from the properties if it's there so that ragged columns will always return the ragged representation.

File: merlin/models/tf/inputs/base.py (hunk 1 above)
Comment (gabrielspmoreira) [reply merlin/models/tf/inputs/base.py]: I changed that. As there are many blocks that don't use `InputBlockV2` and `InputBlock`, I moved the calling of `PrepareFeatures` to `Model`, `Encoder` and to the blocks that can be used as `pre` for `fit() / evaluate()` and the ones that can be used as a transform for `Loader.map()`

File: merlin/models/tf/transforms/features.py [hunk 2]
```diff
@@ -104,23 +105,282 @@ class ToDense(FeaturesTensorTypeConversion):
...
+                                    f"List feature {name} does not have maximum"
+                                    "length set in the schema ({col_schema_shape})"
+                                )
+                            seq_length = int(col_schema_shape.dims[1].max)
+
+                        output_shapes[name] = tf.TensorShape([batch_size, seq_length, 1])
+
+                    else:
+                        if len(input_shapes[name]) == 2:
+                            output_shapes[name] = input_shapes[name] + [1]
```
Comment (karlhigley) [inline merlin/models/tf/transforms/features.py:245]: Based on the error message from the test failures about how only tuples and not lists can be concatenated onto tuples, I think this might need to be:
```suggestion
                            output_shapes[name] = input_shapes[name] + (1,)
```

Comment (gabrielspmoreira) [reply merlin/models/tf/transforms/features.py:245]: Hey @karlhigley . Thanks for the review and suggestion. Typically we can concatenate `tf.TensorShape()` (`input_shapes[name]`) with `list`. I think CI is using the dataloader from main branch, which still returns tuple for list features.
I suspect for some reason the changes I did in [tox.ini](https://github.com/NVIDIA-Merlin/models/pull/999/commits/83f195334578e8fb5e0c7d0c04fe50882d7ca118) are not being able to install the dataloader from [PR #101](https://github.com/NVIDIA-Merlin/dataloader/pull/101).

Comment (karlhigley) [reply merlin/models/tf/transforms/features.py:245]: Curious: Does `TensorShape` concatenate with tuples too, or only lists?

Comment (karlhigley) [reply merlin/models/tf/transforms/features.py:245]: It looks like either one of these should work with `TensorShape` but tuples will only concat other tuples, so I think it's safe to switch over to a tuple here:
```
>>> shape = tf.TensorShape((1,2,3))
>>> shape + [4]
TensorShape([1, 2, 3, 4])
>>> shape + (4,)
TensorShape([1, 2, 3, 4])

>>> shape = (1,2,3)
>>> shape + [4]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: can only concatenate tuple (not "list") to tuple
>>> shape + (4,)
(1, 2, 3, 4)
```

Comment (gabrielspmoreira) [reply merlin/models/tf/transforms/features.py]: Sure @karlhigley . Changed it to tuple as you suggested

File: tox.ini [hunk 3]
```diff
@@ -38,8 +41,10 @@ setenv =
 sitepackages=true
 commands =
     python -m pip install --upgrade git+https://github.com/NVIDIA-Merlin/core.git@{posargs:main}
-    python -m pip install --upgrade git+https://github.com/NVIDIA-Merlin/dataloader.git@{posargs:main}
+    #python -m pip install --upgrade git+https://github.com/NVIDIA-Merlin/dataloader.git@{posargs:main}
+    python -m pip install --upgrade git+https://github.com/bschifferer/dataloader.git@change_output
     python -m pip install --upgrade git+https://github.com/NVIDIA-Merlin/nvtabular.git@{posargs:main}
+    python -m pip install --upgrade git+https://github.com/bschifferer/dataloader.git@change_output
     horovodrun -np 2 sh examples/usecases/multi-gpu/hvd_wrapper.sh pytest -m horovod -rxs tests/unit
```
Comment (edknv) [inline tox.ini:48]: I noticed that the multi-gpu test is failing. Can we try `python -m pytest` here? i.e., `horovodrun -np 2 sh examples/usecases/multi-gpu/hvd_wrapper.sh python -m pytest -m horovod -rxs tests/unit`

Comment (gabrielspmoreira) [reply tox.ini:48]: Thanks @edknv . Let me try that

Comment (karlhigley) [reply tox.ini]: Looks like this hits version conflicts on Tensorflow. Maybe we have different versions installed on the CI container vs in the Tox environment?

Comment (edknv) [reply tox.ini]: PR for bumping the TF version: https://github.com/NVIDIA-Merlin/Merlin/pull/849.

@gabrielspmoreira In the meantime, we could also try removing:
```
deps =
    -rrequirements/test.txt
```
on lines 32-33 to unblock this. Those line will install TF 2.10 in the tox environment, while the ci-runner still has TF 2.9, which seems to be leading to conflict. Maybe worth a try?

Comment (gabrielspmoreira) [reply tox.ini:48]: Just pushed that change to try

Comment (gabrielspmoreira) [reply tox.ini:48]: Yeah @karlhigley . We still have the horovod multi-GPUs tests failing for not using the modified dataloader

Comment (edknv) [reply tox.ini:48]: Sorry I think we should still have `python -m pytest` in the GPU test (AND remove `-rrequirements/test.txt` at the same time). :crossed_fingers:

File: merlin/datasets/testing/sequence_testing/schema.json [hunk 4]
```diff
@@ -26,8 +26,7 @@
       "name": "item_age_days_norm",
       "type": "FLOAT",
       "valueCount": {
-        "min": "1",
-        "max": "4"
```
Comment (oliverholworthy) [inline merlin/datasets/testing/sequence_testing/schema.json:30]: Why do we need to remove the max value count from this schema and others?

File: merlin/models/tf/core/combinators.py [hunk 5]
```diff
@@ -615,7 +615,11 @@ def build(self, input_shape):
     def _maybe_filter_layer_inputs_using_schema(self, name, layer, inputs):
         maybe_schema = getattr(layer, "_schema", None)
         if maybe_schema and isinstance(inputs, dict):
-            layer_inputs = {k: v for k, v in inputs.items() if k in maybe_schema.column_names}
+            layer_inputs = {
+                k: v
+                for k, v in inputs.items()
+                if k.replace("__values", "").replace("__offsets", "") in maybe_schema.column_names
```
Comment (oliverholworthy) [inline merlin/models/tf/core/combinators.py:621]: what situation is this change required for. If we're handling the ragged inputs in the first layer how are the values/offsets making their way through to a ParallelBlock?

File: merlin/models/tf/core/encoder.py [hunk 6]
```diff
@@ -159,11 +166,14 @@ def batch_predict(
 
         return merlin.io.Dataset(predictions)
 
-    def call(self, inputs, training=False, testing=False, targets=None, **kwargs):
+    def call(self, inputs, targets=None, training=False, testing=False, **kwargs):
```
Comment (oliverholworthy) [inline merlin/models/tf/core/encoder.py:169]: is the re-ordering of the params a functional change. e.g. are there any places where these are being passed in as positional params instead of keyword. One option could be to enforce this by adding a `*` after inputs. That way we can be sure that the order of the keyword arguments don't matter. 

```python
def call(self, inputs, *, targets=None, training=False, testing=False, **kwargs):
```

File: merlin/models/tf/models/base.py [hunk 7]
```diff
@@ -142,18 +142,13 @@ def get_output_schema(export_path: str) -> Schema:
...
-            if not is_ragged:
-                properties["value_count"] = {"min": shape[1], "max": shape[1]}
+        if shape.rank > 1 and (shape[1] is not None and shape[1] > 1):
             col_schema = ColumnSchema(
                 output_name,
                 dtype=output_spec.dtype.as_numpy_dtype,
-                is_list=True,
-                is_ragged=is_ragged,
-                properties=properties,
+                dims=((0, None), (shape[1], shape[1])),
```
Comment (oliverholworthy) [inline merlin/models/tf/models/base.py:149]: This can be simplified to `dims=(None, shape[1])`. Equivalent to what is already here.

File: merlin/datasets/testing/sequence_testing/schema.json (hunk 4 above)
Comment (gabrielspmoreira) [reply merlin/datasets/testing/sequence_testing/schema.json:30]: The issue is that when min and max are provided and are different, the dataloader returns a tf.SparseTensor, but now MM can only take either tf.Tensor or our convension for ragged tensors (__values and __offsets). So I remove the max to enforce it to return the ragged tensors.
I know you have already a [draft PR](https://github.com/NVIDIA-Merlin/dataloader/pull/103) that will make it return ragged tensors in such cases. As soon as it is merged we can add back the `value_count.max` from the test dataset schemas if needed.
Does that make sense?

File: merlin/models/tf/core/combinators.py (hunk 5 above)
Comment (gabrielspmoreira) [reply merlin/models/tf/core/combinators.py:621]: You are right. In general we use `PrepareFeatures` in the `Model` right at the beginning to ensure that lists are converted from our ragged tensor convension (__values and __offsets) to `tf.RaggedTensor`.
But for `RetrievalModelV2` we cannot `PrepareFeatures` in `Model`, because in that case the `Encoders` (which are `tf.keras.Model`) would receive `tf.RaggedTensor` as inputs and that would cause problems when the query encoder is saved to be served on Triton.
That is why on `RetrievalModelV2` constructor we set `prep_features=False` and let the `Encoder` class to `prep_features=True`. As the `RetrievalModelV2` is a `ParallelBlock` of `Encoder` that is when it would receive the ragged convension (`__values` and `__offsets`) and this change is to ensure that the ragged representation is cascaded to the `Encoder` block

File: merlin/models/tf/models/base.py (hunk 7 above)
Comment (gabrielspmoreira) [reply merlin/models/tf/models/base.py]: Done

File: merlin/models/tf/core/encoder.py (hunk 6 above)
Comment (gabrielspmoreira) [reply merlin/models/tf/core/encoder.py]: Done

**Comment by bschifferer** [comment]
rerun tests
